lib/elastic_apm/trace_context.rb

```diff
@@ -1,82 +1,107 @@
 # frozen_string_literal: true
 
+require 'elastic_apm/trace_context/tracestate'
+require 'elastic_apm/trace_context/traceparent'
+
 module ElasticAPM
   # @api private
   class TraceContext
     class InvalidTraceparentHeader < StandardError; end
 
-    VERSION = '00'
-    HEX_REGEX = /[^[:xdigit:]]/.freeze
-
-    TRACE_ID_LENGTH = 16
-    ID_LENGTH = 8
-
     def initialize(
-      version: VERSION,
-      trace_id: nil,
-      span_id: nil,
-      id: nil,
-      recorded: true
+      traceparent: nil,
+      tracestate: nil,
+      **legacy_traceparent_attrs
     )
-      @version = version
-      @trace_id = trace_id || hex(TRACE_ID_LENGTH)
-      # TODO: rename to parent_id with next major version bump
-      @parent_id = span_id
-      @id = id || hex(ID_LENGTH)
-      @recorded = recorded
+      @traceparent = traceparent || Traceparent.new(**legacy_traceparent_attrs)
+      @tracestate = tracestate
     end
 
-    attr_accessor :version, :id, :trace_id, :parent_id, :recorded
+    attr_accessor :traceparent, :tracestate
 
-    alias :recorded? :recorded
-    def self.parse(header)
-      raise InvalidTraceparentHeader unless header.length == 55
-      raise InvalidTraceparentHeader unless header[0..1] == VERSION
-
-      new.tap do |t|
-        t.version, t.trace_id, t.parent_id, t.flags =
-          header.split('-').tap do |values|
-            values[-1] = Util.hex_to_bits(values[-1])
-          end
+    def self.parse(legacy_header = nil, env: nil)
+      if !legacy_header && !env
+        raise ArgumentError, 'TraceContext expects either env: or single ' \
+          'argument header string'
+      end
 
-        raise InvalidTraceparentHeader if HEX_REGEX =~ t.trace_id
-        raise InvalidTraceparentHeader if HEX_REGEX =~ t.parent_id
+      if legacy_header
+        return legacy_parse_from_header(legacy_header)
       end
+
+      return unless (header = get_traceparent_header(env))
+
+      parent = TraceContext::Traceparent.parse(header)
+
+      state =
+        if (header = env['HTTP_TRACESTATE'])
+          TraceContext::Tracestate.parse(header)
+        end
+
+      new(traceparent: parent, tracestate: state)
     end
 
-    def flags=(flags)
-      @flags = flags
+    def self.legacy_parse_from_header(header)
+      parent = Traceparent.parse(header)
+      new(traceparent: parent)
+    end
 
-      self.recorded = flags[7] == '1'
+    def version
+      traceparent.version
     end
 
-    def flags
-      format('0000000%d', recorded? ? 1 : 0)
+    def trace_id
+      traceparent.trace_id
     end
 
-    def hex_flags
-      format('%02x', flags.to_i(2))
+    def id
+      traceparent.id
+    end
+
+    def parent_id
+      traceparent.parent_id
     end
 
     def ensure_parent_id
-      @parent_id ||= hex(ID_LENGTH)
+      traceparent.ensure_parent_id
+    end
+
+    def recorded?
+      traceparent.recorded?
```

Comment:
Each of these methods could also be forwarded to `traceparent` to make the class smaller : )